src/analysis/seminorm.lean

```diff
@@ -39,12 +43,130 @@ open_locale big_operators nnreal pointwise topological_space
 
 variables {R R' 𝕜 E F G ι : Type*}
 
+/-- A seminorm on an add_monoid is a function to the reals that preserves zero, takes nonnegative
+values, and is subadditive. -/
+structure add_monoid_seminorm (G : Type*) [add_monoid G]
+  extends zero_hom G ℝ :=
+(nonneg : ∀ r, 0 ≤ to_fun r)
+(add : ∀ r s, to_fun (r + s) ≤ to_fun r + to_fun s)
```

Comment:
Renaming `add` to `add_le'` would be more consistent with the corresponding lemma.